### ISSUE TITLE
Paper #633: Interval Counterexamples for Loop Invariant Learning

### ISSUE THREAD

**Issue by kstolee** [researchart/fse20#66]
This issue is related to PR #57

Authors are applying for the *functional* and *available* badges.

Author contact: @xurongchen

**Comment by bigfatnoob** [comment]
## Authors:
Name	| Email	GitHub ID	Corresponding author
Rongchen Xu xrc19@mails.tsinghua.edu.cn @xurongchen ✔️

**Comment by bigfatnoob** [comment]
@xurongchen : Can you post a comment with the names, emails, and GitHub IDs of the co-authors for your paper?

**Comment by art-reviewer** [comment]
[Paper](https://github.com/xurongchen/fse20/blob/f7950be7fbad6b1625d24df5ea2d1d2fb6a3e13b/submissions/functional/paper%23633/paper%23633.pdf)

**Comment by xurongchen** [comment]
| Author          | Mail                        | GitHub ID      |                      | 
|-----------------|-----------------------------|----------------|----------------------|
| Rongchen Xu     | xrc19@mails.tsinghua.edu.cn | @xurongchen     |                      |
| Fei He          | hefei@tsinghua.edu.cn       | @feihe          | Corresponding Author |
| Bow-Yaw Wang    | bywang@iis.sinaca.edu.tw    |                |                      |

Author contact: @xurongchen

**Comment by anon-eval** [comment]
@xurongchen, @feihe
Thank you for submitting the artifacts. Congratulations on the paper :)

**TLDR**: Based on the current version of your submission, I recommend the **available** and **functional** badges. 


I arrived at this conclusion using the following metrics
1. **Installation**✅ : The installation runs successfully. It would be helpful if you can provide a requirements.txt file with the python requirements rather than just mentioning them. It could lead to version conflicts.
2. **Public Availability**✅ : The tool is publicly available on github. It would also be nice if you could upload the artifacts on [doi.org](doi.org) or [zenodo](https://zenodo.org/)
3. **Documentation** ✅  : The tool is well documented for reproducing the results and executing it.
4. **Consistent/Complete**✅ : The results are consistent on multiple executions and the results in the paper can be reproduced using the scripts provided.
5. **Exercisable** ✅ : The tool is exercisable on the datasets provided in the repository
6. **Evidence of verification and validation** ✅ : I was able to obtain the results in the paper executing the `RunXXX.py` scripts.

Thanks :)

**Comment by letmecheckthat** [comment]
Based on a preliminary check, there are multiple issues with this artifact:

- Absolute/hard-coded paths are used in many Python scripts, which adds a lot of overhead to running these. Please make these relative.
- These paths are also hard-coded in the binaries for the various benchmarks, which is a serious encumbrance, effectively requiring replication efforts to use the exact path specified. If this is intended, please update the INSTALL.md descriptions (which currently just state to "check the paths"), but note that I consider this a substantial obstacle to even a 'functional' badge.
- Related, even assuming relative paths, the main benchmark directory path in the scripts points to a `PInferLoopInvBechmark/All` directory that does not exist; there is a `*repo*/Benchmarks` directory.
- The main executor script (Experiment.py) contains many commented-out, presumably defunct lines of code (as well as comments that state to "check" the correctness of statements, which is worrisome). Please clean this up; it inhibits understandability.
- Regarding **Availability**, this artifact does not appear to be hosted on an archival repository. Please make one. For reference, you can make a Zenodo archive that tracks releases from a Github repository and archives each new release.
- More generally, the absence of virtually all source code pertaining to the tool's implementation (i.e., the VS Solutions) is problematic. The README states that this is due to potential copyright issues, which is understandable; however, I am hesitant to assign any artifact badge without access to the corresponding source code.

**Comment by xurongchen** [comment]
Thanks for your feedback. We made the following modifications based on your suggestions:

* We changed the absolute path in the script to a relative path that matches the current directory structure.
* Our executable binaries do not use hard-coded paths. The python script file that encodes the benchmark path only provides a quick way to perform experiments. For more general use, we have updated INSTALL.md to provide a way to execute a specific use case individually. 
* We cleaned up invalid comments to make the script more readable.
* By contacting the author of previous work, we obtained their open source consent and updated the required open source license: Microsoft Public License (MSPL). All source code content has been added to the repository. 
* We made a Zenodo archive that tracks releases from a Github repository: 
[![DOI](https://zenodo.org/badge/269244501.svg)](https://zenodo.org/badge/latestdoi/269244501)

I have pulled a new request for the above. Looking forward to your follow-up feedback.

**Comment by letmecheckthat** [comment]
Thank you, that's great progress! There seem to be a few more path-related issues; e.g., the relative path `../Benchmark` triggers an error in `Experiment.py` because the working directory is changed in there (also, the folder is named `Benchmark*s*`). Please fix that and check that the code works for you from a fresh checkout; I will resume checking in as soon as it does.

**Comment by xurongchen** [comment]
Thanks. I have fixed the path-related problem and tested the python scripts. Now the scripts run correctly on my machine with current file structure.

**Comment by letmecheckthat** [comment]
Just to check: did you try this in any other directory than `d:\CodeRepository` on your machine (e.g., in a random other sub-directory)? Because this crashes when I try to do so, with a `The system cannot find the file specified` error, mentioning a number of `d:\CodeRepository\` paths in the trace. Running a `grep -r "d:\\\\CodeRepository" Tools/` indeed finds this path hard-coded in (apparently) all DLL's and PDBs in that directory. Please make sure that the paths in those executables are relative too, or at least parameterized so that I can run this in another directory.

**Comment by xurongchen** [comment]
Yes, I used another directory and tested the scripts. Also I renamed my "d:\\CodeRepository" directory to check there are no dependences to the directory "d:\\CodeRepository".  Interestingly, when I run `grep -r "d:\\\\CodeRepository" Tools/`, I also find the problem you mentioned. I checked one DLL file and I found it is in a comment which the compiler added at the compile time. I am sure that there are no hard-coded pathes in DLL's and PDBs.

**Comment by kstolee** [comment]
@xurongchen please tag your pull requests in your comments. I've edited them to include the links.

**Comment by xurongchen** [comment]
Thanks, I will notice that in future.

**Comment by kstolee** [comment]
There seems to be consensus on ```available``` and I'm happy to see the new DOI. 

@letmecheckthat are you able to make a recommendation yet regarding the ```functional``` badge?

**Comment by letmecheckthat** [comment]
Thanks, can confirm that the code is runnable, although the reliance on (severely out-dated) tools like MinGW made it a rather tedious experience. It might be helpful to write every step out in more detail; for instance, provide links to the specific versions of MinGW, MSYS, Cmake; their typical destinations, and common failure modes (e.g., I had to copy two DLLs to the folder containing `IDT4Inv.exe` or set up static linking). In any case, Functional seems appropriate.

**Comment by kstolee** [comment]
@xurongchen please cite your artifact using the DOI in the references section of your paper. 

I also suggest you follow the suggestions by @letmecheckthat on providing more details on every step in the process. There is a lot of wisdom in this discussion that would be helpful to the next person who tries to run your code.

**Comment by xurongchen** [comment]
Thank you for your suggestion, I further updated the installation process (INSTALL.md), especially the installation information of MinGW. New PR: #166
